### PR TITLE
HDDS-2848. Recon changes to make snapshots work with OM HA.

### PR DESCRIPTION
**Pull request by swagle**
## What changes were proposed in this pull request?
Implemented changes proposed in the Jira and added an integration test.

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-2848

## How was this patch tested?
New integration test added.

### CODE REVIEW

**Comment by swagle** [comment]
@avijayanhwx / @bharatviswa504 could you guys please review?

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/recon/TestReconWithOzoneManagerHA.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+    String hostname =
+        ozoneManager.get().getHttpServer().getHttpAddress().getHostName();
+    String expectedUrl = "http://" +
+        (hostname.equals("0.0.0.0") ? "localhost" : hostname) + ":" +
+        ozoneManager.get().getHttpServer().getHttpAddress().getPort() +
+        OZONE_OM_DB_CHECKPOINT_HTTP_ENDPOINT;
+    String snapshotUrl = impl.getOzoneManagerSnapshotUrl();
+    Assert.assertEquals("OM Snapshot should be requested from the leader.",
+        expectedUrl, snapshotUrl);
```
Comment (avijayanhwx) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/recon/TestReconWithOzoneManagerHA.java]: Do we need to trigger an OM DB sync here and see if the sync is working?

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/spi/impl/OzoneManagerServiceProviderImpl.java [hunk 2]
```diff
@@ -237,19 +241,47 @@ public void stop() throws Exception {
     scheduler.shutdownNow();
   }
 
+  /**
+   * Find the OM leader's address to get the snapshot from.
+   */
+  @VisibleForTesting
+  public String getOzoneManagerSnapshotUrl() throws IOException {
+    if (!configuration.getBoolean(
+        OMConfigKeys.OZONE_OM_RATIS_ENABLE_KEY, false)) {
```
Comment (avijayanhwx) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/spi/impl/OzoneManagerServiceProviderImpl.java]: For single node RATIS, we can still use the regular URL.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/recon/TestReconWithOzoneManagerHA.java [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    OzoneConfiguration conf = new OzoneConfiguration();
+    conf.set(OMConfigKeys.OZONE_OM_RATIS_ENABLE_KEY, Boolean.TRUE.toString());
+    cluster = (MiniOzoneHAClusterImpl) MiniOzoneCluster.newHABuilder(conf)
+        .setClusterId(UUID.randomUUID().toString())
+        .setScmId(UUID.randomUUID().toString())
+        .setOMServiceId(omServiceId)
+        .setNumDatanodes(1)
+        .setNumOfOzoneManagers(3)
+        .includeRecon(true)
+        .setReconHttpPort(NetUtils.getFreeSocketPort())
```
[the remark this answers is not in the document]
Comment (swagle) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/recon/TestReconWithOzoneManagerHA.java]: Fair enough.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/recon/TestReconWithOzoneManagerHA.java (hunk 1 above)
Comment (swagle) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/recon/TestReconWithOzoneManagerHA.java]: I intended to test only changes I made but this Test class I envision can easily extend to add more tests for HA but those can be separate Jiras, IMO.

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/spi/impl/OzoneManagerServiceProviderImpl.java (hunk 2 above)
Comment (swagle) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/spi/impl/OzoneManagerServiceProviderImpl.java]: OM AFAIK is either 3 node RATIS or no HA based on this configuration, although not sure why we cannot check, the result should be the same.

**Comment by avijayanhwx** [comment]
`org.apache.hadoop.ozone.freon.TestDataValidateWithUnsafeByteOperations`
Failure seems unrelated.

**Comment by avijayanhwx** [comment]
Thank you for working on this @swagle.

**Pull request merged**